cmd/kube-rbac-proxy/app/options/proxyoptions.go

```diff
@@ -125,13 +139,101 @@ func (o *ProxyOptions) ApplyTo(c *server.KubeRBACProxyInfo, a *serverconfig.Auth
 		}
 	}
 
+	c.UpstreamHeaders = o.UpstreamHeader
 	c.IgnorePaths = o.IgnorePaths
 	c.AllowPaths = o.AllowPaths
 	a.APIAudiences = o.TokenAudiences
 
 	return nil
 }
 
+func validateSecureConnectionConfig(o *ProxyOptions) error {
+	if !identityheaders.HasIdentityHeadersEnabled(o.UpstreamHeader) || !o.UpstreamForceH2C {
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(loopbackLookupTimeout)*time.Second)
+	defer cancel()
+
+	isLoopback, errLoopback := isLoopbackAddress(ctx, o.Upstream)
+	if errLoopback == nil && isLoopback {
+		// If it is a loopback address, the connection is considered secure.
+		return nil
+	}
+
+	if o.UpstreamForceH2C {
+		return fmt.Errorf("loopback address is required for h2c")
+	}
+
+	// Handles the case where the upstream address couldn't be verified as loopback.
+	errMTLS := validateMTLSConfiguration(o.UpstreamClientCertFile, o.UpstreamClientKeyFile, o.UpstreamCAFile)
+	if errMTLS == nil {
+		// A mTLS setup is considered secure.
+		return nil
+	}
+
+	// Handle the case where upstream couldn't be verified as loopback and mTLS is not configured.
+	if errLoopback == nil {
+		errLoopback = errors.New("not a loopback address")
+	}
+
+	return fmt.Errorf(
+		"loopback address or mTLS is required for identity headers: loopback - %w; mTLS - %w",
+		errLoopback,
+		errMTLS,
+	)
+}
+
+func validateMTLSConfiguration(upstreamClientCertPath, upstreamClientKeyPath, upstreamCAPath string) error {
+	// Check if client is configured to provide a certificiate.
+	if len(upstreamClientCertPath) == 0 || len(upstreamClientKeyPath) == 0 || len(upstreamCAPath) == 0 {
+		return fmt.Errorf("krp client cert/key and an upstream CA are required for mTLS")
+	}
+
+	_, err := tls.LoadX509KeyPair(upstreamClientCertPath, upstreamClientKeyPath)
```

Comment:
aren't these validated elsewhere?

Reply:
It is being done in the Config setup. As it is only done on startup, I am accepting it, to keep the code clean.

Though process:
1. Validate is called before Complete, which makes sense. Why should you try to complete on invalid flags? So populating options with it is not feasible.
2. If Validate would fill up the Options / Config, it would violate the single purpose in its name.

WDYT?

Reply:
I don't think the current shape is correct:

You only check the correctness of the upstream CA and upstream client cert/key pair in case we've got identity headers or we're forcing H2C.
You still won't check them in case upstream is a loopback address, even though upstream may require client cert auth.

That means you've got validation that mostly doesn't work.

I see two possible ways to get out of this:
1. you always verify the options
2. you'll delegate the verification of the certs to the `ApplyTo()` code and only validate that the flag combination is correct